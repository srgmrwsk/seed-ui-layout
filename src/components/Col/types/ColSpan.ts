export type ColSpan = boolean | number | 'auto';
