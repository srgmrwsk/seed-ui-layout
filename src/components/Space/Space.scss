.seed-space {
  display: flex;

  &_gutter {
    &._seed-direction_column,
    &._seed-direction_column-reverse {
      > *:not(last-of-type) {
        margin-bottom: var(--seed-space-gutter);
      }
    }

    &._seed-direction_row,
    &._seed-direction_row-reverse {
      > *:not(last-of-type) {
        margin-right: var(--seed-space-gutter);
      }
    }
  }

  &__item {
    line-height: 1;
  }
}
