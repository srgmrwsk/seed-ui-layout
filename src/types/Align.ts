export type Align = 'start' | 'end' | 'center' | 'baseline' | 'stretch';
