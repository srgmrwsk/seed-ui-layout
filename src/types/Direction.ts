export type Direction = 'column' | 'column-reverse' | 'row' | 'rew-reverse';
