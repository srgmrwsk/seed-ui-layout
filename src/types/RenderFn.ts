import React from 'react';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type RenderFn = (component: React.ReactElement, ...args: []) => any;
