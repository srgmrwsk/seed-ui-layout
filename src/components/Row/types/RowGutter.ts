import { Sizing } from '../../../types/Sizing';

export type RowGutter = Sizing | [Sizing, Sizing];
