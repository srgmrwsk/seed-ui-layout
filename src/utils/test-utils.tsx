import React from 'react';
import { render } from '@testing-library/react';
import { renderToStaticMarkup } from 'react-dom/server';
import { configureAxe } from 'jest-axe';

import { RenderFn } from '../types/RenderFn';

export const axe = configureAxe({
  rules: {
    // disable landmark rules when testing isolated components
    region: { enabled: false },
  },
});

export function create(
  ...args: Parameters<RenderFn>
): ChildNode | HTMLCollection | null {
  const { container } = render(...args);
  return container.children.length > 1
    ? container.children
    : container.firstChild;
}
export function renderToHtml(component: React.ReactElement): string {
  return renderToStaticMarkup(component);
}
