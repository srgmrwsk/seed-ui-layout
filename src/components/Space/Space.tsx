import React from 'react';
import cx from 'classnames';
import { isNil } from 'ramda';

import { getSpacingVariables } from '../../utils/get-spacing-variables';
import { Align } from '../../types/Align';
import { Justify } from '../../types/Justify';
import { getStringStyles } from '../../utils/get-string-styles';
import { Direction } from '../../types/Direction';
import { getBooleanStyles } from '../../utils/get-boolean-styles';
import { Sizing } from '../../types/Sizing';

export interface SpaceProps
  extends Omit<React.AllHTMLAttributes<HTMLElement>, 'as'> {
  as?: React.ElementType;
  align?: Align;
  direction?: Direction;
  gutter?: Sizing;
  justify?: Justify;
}

function Space(
  {
    as: As = 'div',
    align,
    className,
    direction,
    gutter,
    justify,
    style,
    children,
    ...elemProps
  }: SpaceProps,
  ref: React.Ref<HTMLElement>,
): JSX.Element {
  return (
    <As
      {...elemProps}
      ref={ref}
      className={cx(
        'seed-space',
        getBooleanStyles('seed-space_gutter', gutter),
        getStringStyles('_seed-align', align),
        getStringStyles('_seed-justify', justify),
        getStringStyles('_seed-direction', direction),
        className,
      )}
      style={{
        ...getSpacingVariables('seed-space-gutter', gutter),
        style,
      }}
    >
      {React.Children.map(children, (child) =>
        isNil(child) ? child : <div className="seed-space__item">{child}</div>,
      )}
    </As>
  );
}

Space.displayName = 'Space';

export default React.forwardRef(Space);
