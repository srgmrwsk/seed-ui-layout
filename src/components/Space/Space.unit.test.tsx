import React from 'react';
import { renderToStaticMarkup } from 'react-dom/server';

import { RenderFn } from '../../types/RenderFn';
import { axe, create } from '../../utils/test-utils';

import Space, { SpaceProps } from './Space';

describe('Space', () => {
  const setup = (renderFn: RenderFn, props?: SpaceProps) =>
    renderFn(<Space {...props} />);

  it('should render correctly with default props', () => {
    const actual = setup(create);
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with as prop', () => {
    const actual = setup(create, { as: 'section' });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with children prop', () => {
    const actual = setup(create, { children: <p>Element with text.</p> });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with children array prop', () => {
    const actual = setup(create, {
      children: [
        <p key="0">Element with text 1.</p>,
        <p key="2">Element with text 2.</p>,
      ],
    });

    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with align prop', () => {
    const actual = setup(create, { align: 'center' });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with direction prop', () => {
    const actual = setup(create, { direction: 'column' });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with gutter prop', () => {
    const actual = setup(create, { gutter: 3 });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with justify prop', () => {
    const actual = setup(create, { justify: 'center' });
    expect(actual).toMatchSnapshot();
  });

  it('should meet accessibility guidelines', async () => {
    const wrapper = setup(renderToStaticMarkup);
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });
});
