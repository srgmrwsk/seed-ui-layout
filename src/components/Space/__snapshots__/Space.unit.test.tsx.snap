// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Space should render correctly with align prop 1`] = `
<div
  class="seed-space _seed-align_center"
/>
`;

exports[`Space should render correctly with as prop 1`] = `
<section
  class="seed-space"
/>
`;

exports[`Space should render correctly with children array prop 1`] = `
<div
  class="seed-space"
>
  <div
    class="seed-space__item"
  >
    <p>
      Element with text 1.
    </p>
  </div>
  <div
    class="seed-space__item"
  >
    <p>
      Element with text 2.
    </p>
  </div>
</div>
`;

exports[`Space should render correctly with children prop 1`] = `
<div
  class="seed-space"
>
  <div
    class="seed-space__item"
  >
    <p>
      Element with text.
    </p>
  </div>
</div>
`;

exports[`Space should render correctly with default props 1`] = `
<div
  class="seed-space"
/>
`;

exports[`Space should render correctly with direction prop 1`] = `
<div
  class="seed-space _seed-direction_column"
/>
`;

exports[`Space should render correctly with gutter prop 1`] = `
<div
  class="seed-space seed-space_gutter"
  style="--seed-space-gutter: 24px;"
/>
`;

exports[`Space should render correctly with justify prop 1`] = `
<div
  class="seed-space _seed-justify_center"
/>
`;
